src/main/scala/viper/gobra/reporting/StatsCollector.scala

```diff
@@ -0,0 +1,405 @@
+  // if the type checking or other processes are done concurrently in the future
+
+  // Stores type info for a task
+  private[reporting] val typeInfos: Map[String, TypeInfo] = TrieMap()
+  // Maps a gobra member name to a gobra member entry
+  private[reporting] val memberMap: Map[String, GobraMemberEntry] = TrieMap()
+  // Maps a viper member name to a gobra member entry
+  private val viperMemberNameGobraMemberMap: Map[String, GobraMemberEntry] = TrieMap()
+
+  // indentation prefix for generated json
+  private val i = "  "
+  case class GobraMemberEntry(info: GobraMemberInfo, viperMembers: Map[String, ViperMemberEntry]) {
+
+    def asJson(p: String = ""): String = {
+      val viperMembersJson = viperMembers.values.map(_.asJson(s"$p$i$i"))
+      val dependencies = this.dependencies().map(entry => s"""$p$i$i"${entry.key}"""")
+
+      s"""$p{
+         |$p$i"id": "${this.key}",
+         |$p$i"pkgId": "${info.pkgId}",
+         |$p$i"pkg": "${info.pkg}",
+         |$p$i"name": "${info.pkg}.${info.memberName}",
+         |$p$i"args": "${info.args}",
+         |$p$i"nodeType": "${info.nodeType}",
+         |$p$i"trusted": ${info.isTrusted},
+         |$p$i"abstract": ${info.isAbstractAndNotImported},
+         |$p$i"viperMembers": [
+            |${viperMembersJson.mkString(",\n")}
+         |$p$i],
+         |$p$i"dependencies": [
+            |${dependencies.mkString(",\n")}
+         |$p$i]
+         |$p}""".stripMargin
+    }
+
+    def dependencies(): Set[GobraMemberEntry] = {
+      this.viperMembers.values
+        .flatMap(viperMember => viperMember.dependencies
+          .flatMap(dep => findGobraMemberByViperMemberName(viperMember.taskName, dep, info.pkgId)))
+        .filter(_.key != this.key).toSet
+    }
+
+    def key: String = gobraMemberKey(info.pkgId, info.memberName, info.args)
+  }
+
+  case class ViperMemberEntry(memberName: String,
+                              taskName: String,
+                              time: Time,
+                              nodeType: ViperNodeType,
+                              dependencies: Set[String],
+                              success: Boolean,
+                              cached: Boolean,
+                              fromImport: Boolean,
+                              hasBody: Boolean,
+                              verified: Boolean) {
+
+    def asJson(p: String = ""): String = {
+      s"""$p{
+         |$p$i"name": "${this.memberName}",
+         |$p$i"taskName": "${this.taskName}",
+         |$p$i"time": ${this.time},
+         |$p$i"nodeType": "${this.nodeType}",
+         |$p$i"success": ${this.success},
+         |$p$i"cached": ${this.cached},
+         |$p$i"fromImport": ${this.fromImport},
+         |$p$i"hasBody": ${this.hasBody},
+         |$p$i"verified": ${this.verified}
+         |$p}""".stripMargin
+    }
+  }
+
+  case class GobraMemberInfo(pkgId: String,
```

Comment:
why do you replicate parts of the information that we store in the AST? Couldn't you simply store the internal or parse AST node?

Reply:
I worry about memory issues when storing the ASTs, particularly viper ASTs because of all the subnodes that then don't get garbage collected